src/internal/clusterstate/v2.5.0/collection.go

```diff
@@ -267,32 +266,84 @@ func migratePostgreSQLCollection(ctx context.Context, tx *pachsql.Tx, name strin
-	i := 0
+
+	// Optimized code for building the UPDATEs and sending them to Postgres in batches
+	// NOTE: This code necessitated upgrading from sqlx 1.2.0 to sqlx 1.3.x to get batching support for named variables
+
+	log.Info(ctx, fmt.Sprintf("Migrating %s [%d rows]", name, len(vals)))
+
+	batch_size := 1000 // Maximum size of each batch
+	batch_count := 0   // Incrementing count of the number of rows in the current batch
+	total_count := 0   // Incrementing count of the total number of rows
+	query := ""        // String to hold the UPDATE query
+
+	// Create an array that will contain the substring of the SET clause for each column in the UPDATE statement.
+	// For example, the substring for the "idx_name" column would be "idx_name = :idx_name"
+	// In that substring, "idx_name" is the column name, and ":idx_name" is a placeholder for the actual value.
+	columnSubstrings := []string{}
+
+	// Loop through each row that needs to be updated
 	for oldKey, pair := range vals {
+
+		// These two function calls handle updating an oldKey (with no project reference) to a newKey (with project reference)
 		if err := col.withKey(oldKey, func(oldKey string) error {
 			return col.withKey(pair.key, func(newKey string) error {
-				params, err := col.getWriteParams(newKey, pair.val)
+
+				// Increment our counters
+				batch_count++
+				total_count++
+
+				// Get the parameters for the current row
+				params, err := col.getWriteParams(pair.key, pair.val)
+
 				if err != nil {
 					return err
 				}
 
-				updateFields := []string{}
-				for k := range params {
-					updateFields = append(updateFields, fmt.Sprintf("%s = :%s", k, k))
+				// Create the query string. This only needs to be done on the first time through, as only the values change with each row,
+				// not the query string itself.  The query string has placeholders for the values that change from one row to the next.
+				if total_count == 1 {
+
+					// Loop through each parameter in the current row
+					for k := range params {
+						// Add the UPDATE/SET substring for this column
+						columnSubstrings = append(columnSubstrings, fmt.Sprintf("%s = :%s", k, k))
+					}
+
+					query = fmt.Sprintf("update collections.%s set %s where key = :oldKey", col.table, strings.Join(columnSubstrings, ", "))
 				}
+
+				// Add the old key to the parameters. There isn't a column for it (which is why it isn't in params[] to begin with), by it
+				// gets used in the WHERE clause.
 				params["oldKey"] = oldKey
 
-				query := fmt.Sprintf("update collections.%s set %s where key = :oldKey", col.table, strings.Join(updateFields, ", "))
+				// Add the params object to the batch of parameters that we are building
+				param_batch := []any{}
+				param_batch = append(param_batch, params)
+
+				// If we've reached the batch limit, or if we're on the last row, send the updates to Postgres
+				if batch_count == batch_size || total_count == len(vals) {
```

Comment:
```suggestion
				if total_count == len(vals) || total_count%batch_size == 0
```
so that you don't need a separate `batch_count`